File: core/models/quiz_question.py
from __future__ import annotations

class MultipleChoiceQuestion():

    class Choice():
        def __init__(self, description: str, is_correct: bool):
            self.description = description
            self.is_correct = is_correct

        def export_json(self) -> dict:
            json_result = {}
            json_result["description"] = self.description
            json_result["is_correct"] = self.is_correct
            
            return json_result

    def __init__(self,
        question: str = None,
        choices: list[Choice] = None
    ):
        self.question = question
        self.choices: list[MultipleChoiceQuestion.Choice] = choices
    
    def export_json(self) -> dict:
        print("processing")
        json_result = {}

        json_result["question"] = self.question
        json_result["choices"] = []
        for choice in self.choices:
            json_result["choices"].append(choice.export_json())

        return json_result

    @staticmethod
    def process_from_json(json_data: dict) -> MultipleChoiceQuestion:
        # Process question
        question = json_data["question"]

        # Process choices
        json_choices = json_data["choices"]
        choices: list[MultipleChoiceQuestion.Choice] = []

        for json_choice in json_choices:
            description = json_choice["description"]
            is_correct = json_choice["is_correct"]
            new_choice = MultipleChoiceQuestion.Choice(description, is_correct)
            choices.append(new_choice)

        return MultipleChoiceQuestion(question, choices)
    
    @staticmethod
    def process_from_json_lists(json_lists: list[dict]) -> list[MultipleChoiceQuestion]:
        mcq_list: list[MultipleChoiceQuestion] = []
        for json_data in json_lists:
            new_mcq = MultipleChoiceQuestion.process_from_json(json_data)
            mcq_list.append(new_mcq)
        
        return mcq_list

File: src/lib/models/quiz-question.ts
export class Choice {
  description: string;
  isCorrect: boolean;

  constructor(description: string, isCorrect: boolean) {
    this.description = description;
    this.isCorrect = isCorrect;
  }

  exportJson(): { description: string; is_correct: boolean } {
    return {
      description: this.description,
      is_correct: this.isCorrect,
    };
  }

  static fromJson(json: { description: string; is_correct: boolean }): Choice {
    return new Choice(json.description, json.is_correct);
  }
}

export class MultipleChoiceQuestion {
  question: string;
  choices: Choice[];

  constructor(question: string, choices: Choice[]) {
    this.question = question;
    this.choices = choices;
  }

  exportJson(): {
    question: string;
    choices: { description: string; is_correct: boolean }[];
  } {
    return {
      question: this.question,
      choices: this.choices.map((choice) => choice.exportJson()),
    };
  }

  static fromJson(jsonData: {
    question: string;
    choices: { description: string; is_correct: boolean }[];
  }): MultipleChoiceQuestion {
    const question = jsonData.question;
    const choices = jsonData.choices.map((choiceData) =>
      Choice.fromJson(choiceData)
    );
    return new MultipleChoiceQuestion(question, choices);
  }

  static fromJsonList(
    jsonList: {
      question: string;
      choices: { description: string; is_correct: boolean }[];
    }[]
  ): MultipleChoiceQuestion[] {
    return jsonList.map((jsonData) =>
      MultipleChoiceQuestion.fromJson(jsonData)
    );
  }
}
